libyul/optimiser/UnusedFunctionReturnParameterPruner.cpp

```diff
@@ -0,0 +1,193 @@
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libyul/optimiser/UnusedFunctionReturnParameterPruner.h>
+#include <libyul/optimiser/UnusedFunctionsCommon.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/optimiser/ASTWalker.h>
+
+#include <libyul/AsmDataForward.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+
+#include <libsolutil/CommonData.h>
+#include <variant>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::langutil;
+using namespace solidity::yul::unusedFunctionsCommon;
+
+/**
+ * Step 1 of UnusedFunctionReturnParameterPruner: Find functions whose return parameters are not
+ * used in the code, i.e.,
+ *
+ * Look at all VariableDeclaration `let x_1, ..., x_i, ..., x_n := f(y_1, ..., y_m)` and check if
+ * `x_i` is unused. If all function calls to `f` has its i-th return parameter unused, we will mark
+ * its i-th index so that it can be removed in later steps.
+ */
+class FindFunctionsWithUnusedReturns: public ASTWalker
+{
+public:
+	explicit FindFunctionsWithUnusedReturns(
+		Dialect const& _dialect,
+		Block const& _block,
+		map<YulString, size_t> const& _references
+	):
+		m_references(_references),
+		m_dialect(_dialect)
+	{
+		(*this)(_block);
+	}
+	using ASTWalker::operator();
+	void operator()(VariableDeclaration const& _v) override;
+
+	map<YulString, vector<bool>> functions();
+
+private:
+	/// Function name and a boolean mask, where `false` at index `i` indicates that the function
+	/// return-parameter at index `i` in `FunctionDefinition::parameter` is unused at every function
+	/// call-site.
+	map<YulString, vector<bool>> m_callSiteMasks;
+	/// A count of all references to all Declarations.
+	map<YulString, size_t> const& m_references;
+	Dialect const& m_dialect;
+};
+
+/// Find functions whose return parameters are unused. Assuming that the code is in SSA form and
+/// that ForLoopConditionIntoBody, ExpressionSplitter were run, all FunctionCalls with at least one
+/// return value will have the form `let x_1, ..., x_n := f(y_1, ..., y_m)`
+void FindFunctionsWithUnusedReturns::operator()(VariableDeclaration const& _v)
+{
+	if (holds_alternative<FunctionCall>(*_v.value))
+	{
+		YulString const& name = (get<FunctionCall>(*_v.value)).functionName.name;
+
+		if (m_dialect.builtin(name))
+			return;
+
+		if (!m_callSiteMasks.count(name))
+			m_callSiteMasks[name].resize(_v.variables.size(), false);
+
+		for (size_t i = 0; i < _v.variables.size(); ++i)
+			m_callSiteMasks.at(name)[i] =
+				m_callSiteMasks.at(name)[i] || m_references.count(_v.variables[i].name);
+	}
+}
+
+/// Returns a map with function name as key and vector of bools such that if the index `i` of the
+/// vector is false, then the function's `i`-th return parameter is unused at *all* callsites.
+map<YulString, vector<bool>> FindFunctionsWithUnusedReturns::functions()
```

Comment:
Maybe turn this into a static function and make the constructor private? We don't want to have this run twice if we can avoid it.